### PR TITLE
ci(doxygen): add automated Doxygen comment completeness check for public API

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Problem

LVGL's [coding guidelines](https://docs.lvgl.io/master/contributing/pull_requests.html) and [Copilot review instructions](https://github.com/lvgl/lvgl/blob/master/.github/copilot-instructions.md) require all public `lv_*` functions in `.h` files to have Doxygen comments with `@param` and `@return` tags. Today this is enforced entirely through manual code review, which creates two problems:

1. **Reviewer burden** — Maintainers and AI review bots (Copilot, Cubic) repeatedly flag the same class of issue: missing `@param`, incorrect `@return`, absent Doxygen blocks. This generates review round-trips that delay merges.
2. **Inconsistent coverage** — Without automated enforcement, undocumented functions accumulate over time. The existing `compile_docs.yml` workflow only checks that Doxygen XML *generation* succeeds and the docs site *builds* — it does not validate that individual functions have complete `@param`/`@return` tags.

#### Current Codebase State

We ran the new checker against all `src/**/*.h` files to quantify the existing documentation debt:

| Metric | Value |
|--------|-------|
| Total `.h` files scanned | 577 |
| Files checked (excluding libs/gen/conf) | 400 |
| Total `lv_*` function declarations | 2,505 |
| Functions with Doxygen block | 2,065 (82.4%) |
| Functions fully compliant (`@param` + `@return`) | 1,929 (77.0%) |
| Total issues found | 670 |

#### Issue Breakdown

| Category | Count | Share |
|----------|-------|-------|
| Missing Doxygen comment block entirely | 440 | 65.7% |
| Missing `@param` for one or more parameters | 148 | 22.1% |
| Missing `@return` for non-void functions | 82 | 12.2% |

#### Top 10 Files by Issue Count

| Issues | File |
|--------|------|
| 36 | `src/draw/vg_lite/lv_vg_lite_utils.h` |
| 25 | `src/draw/vg_lite/lv_vg_lite_path.h` |
| 24 | `src/draw/sw/blend/neon/lv_draw_sw_blend_neon_to_rgb888.h` |
| 22 | `src/draw/sw/blend/neon/lv_draw_sw_blend_neon_to_rgb565.h` |
| 20 | `src/drivers/wayland/lv_wayland_private.h` |
| 20 | `src/stdlib/builtin/lv_tlsf.h` |
| 19 | `src/draw/lv_draw_buf.h` |
| 17 | `src/draw/sw/blend/riscv_v/lv_draw_sw_blend_riscv_v_to_rgb888.h` |
| 17 | `src/drivers/opengles/opengl_shader/lv_opengl_shader_internal.h` |
| 17 | `src/drivers/sdl/lv_sdl_private.h` |

#### PR Review Activity (GitHub Search)

A search of the lvgl/lvgl repository's PR comments reveals significant reviewer effort spent on Doxygen issues:

| Search Query (in PR comments) | PRs Matched |
|-------------------------------|-------------|
| `doxygen` | 63 |
| `"Doxygen" "incorrect"` | 35 |
| `"@return" "missing"` | 195 |
| `"@param" "missing"` | 73 |
| `"missing documentation"` | 23 |
| `"add doxygen"` | 14 |
| `"please add" "documentation"` | 77 |

These are not unique — a single PR may match multiple queries — but they demonstrate that Doxygen completeness is a recurring review topic across hundreds of PRs.

#### Representative Review Comments (from recent PRs)

- **PR #9937** — Copilot: *"Doxygen for `lv_theme_get_user_data()` refers to a 'pointer to an object'... Please fix the `@param`/`@return` wording"*
- **PR #9583** — Cubic: *"P2: Doxygen return description is incorrect — `lv_msgbox_add_text_fmt` returns a label, not a button, which misleads API docs"*
- **PR #9838** — Copilot: *"The new Doxygen comments say 'a scales ...' which is grammatically incorrect. Please update to 'a scale's ...'"*
- **PR #9400** — Contributor: *"I also added some requests to get Doxygen to add contained documentation to the API documentation page"*

### Solution

This PR adds a lightweight Python script (`scripts/check_doxygen.py`) that validates Doxygen comment completeness for public `lv_*` function declarations in header files, integrated into CI as an incremental PR check.

### How It Works

The checker parses `.h` files using regex-based C function declaration matching and verifies:
- Every `lv_*` function declaration has a preceding `/** ... */` Doxygen block
- Every function parameter has a corresponding `@param` tag
- Every non-void return type has a `@return` tag

**Incremental by design**: In CI, only functions on lines touched by the PR diff are checked. This means:
- Existing undocumented functions (the 670 legacy issues) are **not** flagged
- Only newly added or modified function declarations must be compliant
- Zero noise for contributors working on unrelated code

### Changes

**New file: `scripts/check_doxygen.py`**
- `--diff <range>`: Check only functions on lines changed in a git diff (CI mode)
- `--file <path>`: Check all functions in a specific file
- `--all`: Full scan of all `src/**/*.h` (useful for tracking debt reduction)
- `--self-test`: 37 built-in test cases covering parsing, validation, edge cases, and self-lint
- `--timeout <sec>`: Per-file timeout protection (default: 5s) to prevent CI hangs
- `--verbose`: Progress logging for every file
- Skips auto-generated files (`*_gen.h`), third-party code (`src/libs/`), and config headers
- Handles multi-line declarations, function pointer parameters, `[in]`/`[out]` qualifiers
- Performance: scans all 577 headers in ~0.4s

**New file: `.github/workflows/check_doxygen.yml`**
- Triggers on `pull_request` events only
- Runs self-test first to verify the checker itself is sound
- Then checks only the diff between merge-base and PR head
- Follows the same `concurrency` and `if` patterns as existing CI workflows

### Why Not Extend `compile_docs.yml`?

The existing docs workflow installs Doxygen, Emscripten, Node.js, and builds the entire documentation site. It takes minutes and is designed to verify the docs *build*, not source code *compliance*. This new check:
- Runs in seconds (Python only, no external dependencies)
- Gives precise, actionable error messages with file:line references
- Is incremental (only checks the diff)
- Fails fast with clear guidance on what to fix

### Testing

```bash
# Run the full self-test suite (37 cases)
python3 scripts/check_doxygen.py --self-test

# Check a specific file
python3 scripts/check_doxygen.py --file src/core/lv_obj.h

# Full scan (informational — shows all 670 legacy issues)
python3 scripts/check_doxygen.py --all

# Simulate CI diff check
python3 scripts/check_doxygen.py --diff "HEAD~3...HEAD"
```

### Expected Impact

- **For contributors**: Immediate feedback on missing Doxygen tags before review, with clear error messages and format examples
- **For reviewers**: Eliminates the most common documentation review comment ("please add `@param`/`@return`"), freeing review time for logic and design
- **For the project**: Prevents documentation debt from growing; new code enters the codebase fully documented

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: scripts/check_doxygen.py [hunk 1]
```diff
@@ -0,0 +1,1059 @@
...
+    """Find all .h files under root/src/."""
+    headers = []
+    src_dir = os.path.join(root, "src")
+    if not os.path.isdir(src_dir):
+        return headers
+    for dirpath, _, filenames in os.walk(src_dir):
+        for fn in filenames:
+            if fn.endswith(".h"):
+                headers.append(os.path.join(dirpath, fn))
+    return sorted(headers)
```
Comment (AndreCostaaa) [inline scripts/check_doxygen.py:515]: With the new changes, we now need to get the headers inside `include/lvgl` and not under `src`

**Review by AndreCostaaa — CHANGES_REQUESTED** [review]

File: scripts/check_doxygen.py (hunk 1 above)
Comment (FASTSHIFT) [reply scripts/check_doxygen.py:515]: Done

**Comment by FASTSHIFT** [comment]
cc @AndreCostaaa

**Comment by FASTSHIFT** [comment]
A more stringent comment format check will be added later.
